src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,342 @@
+/**
+ * @file ppo.hpp
+ * @author Xiaohong Ji
+ *
+ * This file is the implementation of PPO class, which implements
+ * proximal policy optimization algorithms.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_PPO_IMPL_HPP
+#define MLPACK_METHODS_RL_PPO_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "ppo.hpp"
+
+namespace mlpack {
+namespace rl {
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::PPO(TrainingConfig config,
+       ActorNetworkType actor,
+       CriticNetworkType critic,
+       PolicyType policy,
+       ReplayType replayMethod,
+       UpdaterType updater,
+       EnvironmentType environment):
+  config(std::move(config)),
+  actorNetwork(std::move(actor)),
+  criticNetwork(std::move(critic)),
+  actorUpdater(std::move(updater)),
+  #if ENS_VERSION_MAJOR >= 2
+  actorUpdatePolicy(NULL),
+  #endif
+  criticUpdater(std::move(updater)),
+  #if ENS_VERSION_MAJOR >= 2
+  criticUpdatePolicy(NULL),
+  #endif
+  policy(std::move(policy)),
+  replayMethod(std::move(replayMethod)),
+  environment(std::move(environment))
```

Comment:
I guess, we are never able to actually train the agent as `deterministic` is not initialized, so it takes a garbage value(not False or 0). Also, `totalSteps` is not initialized

Reply:
it seems that it is. I will look at it this problem.

Reply:
Actually, I have been working on this PR for a few days now, and have made the required changes including removal of some merge conflicts and other bugs, a PR has been made to this branch with the changes..
Also, I have made changes to the backward gradient, please have a look when free..

Lastly, it would be great if you could provide push access to this branch because I am eager to have my recent changes reviewed by you and the other members! With these changes, PPO finally converges for MountainCar and Pendulum perfectly! As for LunarLander, I guess, we'd need to have a look at the implementation, or tune some hyperparameters..

Reply:
> Actually, I have been working on this PR for a few days now, and have made the required changes including removal of some merge conflicts and other bugs, a PR has been made to this branch with the changes..
> Also, I have made changes to the backward gradient, please have a look when free..
> 
> Lastly, it would be great if you could provide push access to this branch because I am eager to have my recent changes reviewed by you and the other members! With these changes, PPO finally converges for MountainCar and Pendulum perfectly! As for LunarLander, I guess, we'd need to have a look at the implementation, or tune some hyperparameters..

Great, I will give access to you now. For the Lunarlander, the implementation is also need to think twice.